Yoast/wordpress-seo
# Re-send tracking data after WordPress core update

## Summary

This PR can be summarized in the following changelog entry:

* Makes sure to re-send tracking data after a WordPress core update.

## Relevant technical choices:

* uses `upgrader_process_complete` instead of `upgrader_post_install`
* prevents duplicate events by using `wp_next_scheduled()`
* improves documentation of the `wpseo_enable_tracking` filter which was initially added on https://github.com/Yoast/wordpress-seo/commit/c919a6d8a07f5a498aad975cf850881b60e70370 then "inverted" on https://github.com/Yoast/wordpress-seo/pull/13523/commits/c776a6dcdb480048b3a41c4ca62239ebbb5d1b9f but the filter documentation wasn't updated. see https://github.com/Yoast/wordpress-seo/pull/13523

## Test instructions
Testing this PR is not that easy.

**Preliminary steps:**
- ideally, this should be merged on Premium and tested there
- alternatively, you need to enable tracking on Free by using this filter somewhere in your code: `add_filter( 'wpseo_enable_tracking', '__return_true', 1 );`
- ideally, this should be tested on a WordPress core update
- either make a dedicated installation with an older WP version and then upgrade
- or (only for testing purposes) you can use a _plugin_ update by changing this line
`if ( ! $upgrader || ! isset( $data['type'] ) || $data['type'] !== 'core' ) {`
to 
`if ( ! $upgrader || ! isset( $data['type'] ) || $data['type'] !== 'plugin' ) {`
then change the version number of a couple plugins and test updating the plugins
- install and activate Debug Bar
- install and activate Debug Bar Cron
- if you need to delete a scheduled cron event, you can use WP CLI:
  - `vagrant ssh`
  - `cd /srv/www/wordpress-develop/public_html/`
  - `wp cron event delete wpseo_send_tracking_data_after_core_update`

**Testing:**
- after a core update, navigate to another admin page
- check the debug bar at the top right: Debug > Cron
- check there's a `wpseo_send_tracking_data_after_core_update` event scheduled in 6 hours
- ideally, you should now trigger a new core update... not sure how to do that
- after a new core update, check Debug > Cron again and verify there's not a new `wpseo_send_tracking_data_after_core_update` event added: only the previous one should be there
- wait 6 hours 😆 and check the data are actually sent

[Edit] together with @johannadevos, we found a simpler way for testing:
**How to trigger a core update:**
- you can simply use a stable version of WordPress or the VVV **default** site (not the develop one!)
- clone and build Yoast SEO there, using this branch of course
- install and activate Debug Bar, and Debug Bar Cron
- go to the WordPress "Updates" admin page
- click "Re-install Now" and wait for the reinstall process to complete
- then navigate to another page and check Debug Bar Cron: you should see the `wpseo_send_tracking_data_after_core_update` event added there, scheduled after 6 hours
- please notice that by default `wp_schedule_single_event()` prevents the same event to be added in the following 10 minutes after it was added for the first time
- thus, you need to wait at least 10 minutes
- then, re-install WordPress again
- after the update, navigate to another page
- check Debug > Cron and verify there's only the previous `wpseo_send_tracking_data_after_core_update` event added
- for no reason, ever, you should see two or more `wpseo_send_tracking_data_after_core_update` events scheduled

In other words, what you see in the screenshot below should _not_ happen:

<img width="1190" alt="Screenshot 2019-11-20 at 18 38 50" src="https://user-images.githubusercontent.com/1682452/69345849-1ff24e80-0c72-11ea-9d48-ddf6efad7619.png">



**If you had success testing with a core update following the instructions above, you can skip this part.**
Testing with plugins update instead of core updates:
- delete all scheduled `wpseo_send_tracking_data_after_core_update` events
- make sure you changed the relevant code to `if ( ! $upgrader || ! isset( $data['type'] ) || $data['type'] !== 'plugin' ) {`
- downgrade the version number of a couple installed plugins
- you can do this with the built-in WordPress Plugin editor
- I used the plugins "Core Control" and "RTL switcher" for this purpose
- go to the WordPress Updates page (it doesn't work if you update a plugin from the `Plugins` page)
- update only one of the plugins
- after the update, navigate to another page
- check Debug > Cron and see the event has been scheduled


## UI changes

## Documentation

## Quality assurance


Fixes https://github.com/Yoast/wordpress-seo/issues/13846
